File: _devserver.mjs
#!/usr/bin/env node
/*
 * Local dev server (JS / Node) — mirrors the production clean-URL rules.
 * ---------------------------------------------------------------------
 * `python -m http.server` (and any plain static server) only serves paths that
 * map to a real file, so pretty routes like /bitcoin-puzzle/71/random or the
 * virtual /bitcoin-dormant/50 fail during LOCAL preview. This replays the same
 * rewrite rules the site's .htaccess (cPanel) and 404.html router (GitHub Pages)
 * use, so `node _devserver.mjs` previews the site exactly like the live host.
 *
 * IMPORTANT: this is a LOCAL PREVIEW TOOL ONLY. Nothing server-side (Node OR
 * Python) runs on the deployed static site. On the live site the SAME clean-URL
 * behavior is provided entirely by BROWSER JavaScript that already ships:
 *     - assets/js/clean-urls.js  (keeps the address bar pretty)
 *     - 404.html                 (client-side router for the virtual routes)
 * So you never deploy this file; it just makes your local machine behave like
 * GitHub Pages / cPanel.
 *
 *     node _devserver.mjs           # serves this folder on 127.0.0.1:8000
 *     node _devserver.mjs 8080      # custom port
 */
import http from 'node:http';
import { readFile } from 'node:fs/promises';
import { statSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.dirname(fileURLToPath(import.meta.url));
const PORT = Number(process.argv[2]) || 8000;

const MIME = {
  '.html': 'text/html; charset=utf-8', '.css': 'text/css; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8', '.mjs': 'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8', '.wasm': 'application/wasm',
  '.svg': 'image/svg+xml', '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
  '.gif': 'image/gif', '.webp': 'image/webp', '.ico': 'image/x-icon',
  '.woff': 'font/woff', '.woff2': 'font/woff2', '.txt': 'text/plain; charset=utf-8',
  '.webmanifest': 'application/manifest+json', '.map': 'application/json',
};

const abs = (rel) => path.join(ROOT, rel.replace(/\//g, path.sep));
const isFile = (rel) => { try { return statSync(abs(rel)).isFile(); } catch { return false; } };
const isDir = (rel) => { try { return statSync(abs(rel)).isDirectory(); } catch { return false; } };

// URL path (decoded, no query) -> on-disk relative path, or null (404).
function resolvePath(p) {
  p = p.replace(/^\/+|\/+$/g, '');
  if (p === '') return isFile('index.html') ? 'index.html' : null;
  if (isFile(p)) return p;                                    // real file
  if (isDir(p) && isFile(p + '/index.html')) return p + '/index.html';

  // 1) generic "serve .html without extension" (/foo/bar -> /foo/bar.html)
  if (isFile(p + '.html')) return p + '.html';

  let m;
  // 2) /key/<value> -> key.html
  if (/^key\/[^/]+$/.test(p) && isFile('key.html')) return 'key.html';
  // 3) /brainwallet/<chain>/<n> -> /brainwallet/<chain>/2.html
  if ((m = p.match(/^brainwallet\/([a-z0-9-]+)\/([0-9]+)$/)) && isFile(`brainwallet/${m[1]}/2.html`)) return `brainwallet/${m[1]}/2.html`;
  // 4) /brainwallet/<n> -> /brainwallet/bitcoin/<n>.html
  if ((m = p.match(/^brainwallet\/([0-9]+)$/)) && isFile(`brainwallet/bitcoin/${m[1]}.html`)) return `brainwallet/bitcoin/${m[1]}.html`;
  // 5) /bitcoin-dormant/address/<addr> -> /bitcoin-dormant/address.html
  if (/^bitcoin-dormant\/address\/[^/]+$/.test(p) && isFile('bitcoin-dormant/address.html')) return 'bitcoin-dormant/address.html';
  // 6) /bitcoin-dormant/<n> -> /bitcoin-dormant/_page.html
  if (/^bitcoin-dormant\/[0-9]+$/.test(p) && isFile('bitcoin-dormant/_page.html')) return 'bitcoin-dormant/_page.html';
  // 7) /<a>/<b> -> /<a>/<b>.html   (e.g. /private-keys/bitcoin)
  if ((m = p.match(/^([a-z0-9-]+)\/([a-z0-9-]+)$/)) && isFile(`${m[1]}/${m[2]}.html`)) return `${m[1]}/${m[2]}.html`;
  // 8) /<a>/<n> -> /<a>/<n>.html   (e.g. /bitcoin-puzzle/42)
  if ((m = p.match(/^([a-z0-9-]+)\/([0-9]+)$/)) && isFile(`${m[1]}/${m[2]}.html`)) return `${m[1]}/${m[2]}.html`;
  // 9) /milk-sad/<chain>/<n> -> /milk-sad/<chain>/1.html
  if ((m = p.match(/^milk-sad\/([a-z0-9-]+)\/([0-9]+)$/)) && isFile(`milk-sad/${m[1]}/1.html`)) return `milk-sad/${m[1]}/1.html`;

  return null;
}

const server = http.createServer(async (req, res) => {
  try {
    const u = new URL(req.url, 'http://127.0.0.1');
    const decoded = decodeURIComponent(u.pathname);
    let target = resolvePath(decoded);
    let status = 200;
    if (!target) { target = isFile('index.html') ? 'index.html' : null; status = 404; } // ErrorDocument fallback
    if (!target) { res.writeHead(404); res.end('Not found'); return; }
    const data = await readFile(abs(target));
    const ext = path.extname(target).toLowerCase();
    res.writeHead(status, {
      'Content-Type': MIME[ext] || 'application/octet-stream',
      'Access-Control-Allow-Origin': '*',
    });
    res.end(data);
  } catch {
    res.writeHead(500); res.end('Server error');
  }
});

server.listen(PORT, '127.0.0.1', () => {
  console.log(`Clean-URL dev server (JS/Node, mirrors .htaccess) -> http://127.0.0.1:${PORT}/`);
  console.log(`Try: http://127.0.0.1:${PORT}/bitcoin-puzzle/71/random`);
  console.log('NOTE: local preview only — the live site needs no server (browser JS handles clean URLs).');
});

File: _devserver.py
#!/usr/bin/env python3
"""
Local dev server that mirrors the production .htaccess clean-URL rules.

`python -m http.server` is a dumb static server: it only serves paths that map
to a real file on disk, so pretty routes like /bitcoin-puzzle/71/random 404
locally even though production (GitHub Pages + cPanel .htaccess) serves them.

This server replays the same rewrite rules the .htaccess uses, so a local
preview at http://127.0.0.1:8000 behaves exactly like the deployed site.
It changes ZERO website files — it only affects how requests are routed locally.

    python _devserver.py            # serves this folder on 127.0.0.1:8000
    python _devserver.py 8080       # custom port
"""
import os
import re
import sys
import urllib.parse
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.dirname(os.path.abspath(__file__))


def _abs(rel):
    return os.path.join(ROOT, rel.replace('/', os.sep))


def _isfile(rel):
    return os.path.isfile(_abs(rel))


def _isdir(rel):
    return os.path.isdir(_abs(rel))


def resolve(path):
    """URL path (decoded, no query) -> on-disk relative path, or None (404)."""
    p = path.strip('/')

    # Root and real files/dirs serve themselves.
    if p == '':
        return 'index.html' if _isfile('index.html') else None
    if _isfile(p):
        return p
    if _isdir(p) and _isfile(p + '/index.html'):
        return p + '/index.html'

    # 1) Generic "serve .html without extension"  (/foo/bar -> /foo/bar.html)
    #    Covers /bitcoin-puzzle/71/random -> /bitcoin-puzzle/71/random.html
    if _isfile(p + '.html'):
        return p + '.html'

    # 2) /key/<value>              -> /key.html   (template renders from path)
    if re.match(r'^key/[^/]+$', p) and _isfile('key.html'):
        return 'key.html'

    # 3) /brainwallet/<chain>/<n>  -> /brainwallet/<chain>/2.html
    m = re.match(r'^brainwallet/([a-z0-9-]+)/([0-9]+)$', p)
    if m and _isfile('brainwallet/%s/2.html' % m.group(1)):
        return 'brainwallet/%s/2.html' % m.group(1)

    # 4) /brainwallet/<n>          -> /brainwallet/bitcoin/<n>.html
    m = re.match(r'^brainwallet/([0-9]+)$', p)
    if m and _isfile('brainwallet/bitcoin/%s.html' % m.group(1)):
        return 'brainwallet/bitcoin/%s.html' % m.group(1)

    # 5) /bitcoin-dormant/address/<addr> -> /bitcoin-dormant/address.html
    if re.match(r'^bitcoin-dormant/address/[^/]+$', p) and _isfile('bitcoin-dormant/address.html'):
        return 'bitcoin-dormant/address.html'

    # 6) /bitcoin-dormant/<n>      -> /bitcoin-dormant/_page.html
    if re.match(r'^bitcoin-dormant/[0-9]+$', p) and _isfile('bitcoin-dormant/_page.html'):
        return 'bitcoin-dormant/_page.html'

    # 7) /<a>/<b>                  -> /<a>/<b>.html   (e.g. /private-keys/bitcoin)
    m = re.match(r'^([a-z0-9-]+)/([a-z0-9-]+)$', p)
    if m and _isfile('%s/%s.html' % (m.group(1), m.group(2))):
        return '%s/%s.html' % (m.group(1), m.group(2))

    # 8) /<a>/<n>                  -> /<a>/<n>.html   (e.g. /bitcoin-puzzle/42)
    m = re.match(r'^([a-z0-9-]+)/([0-9]+)$', p)
    if m and _isfile('%s/%s.html' % (m.group(1), m.group(2))):
        return '%s/%s.html' % (m.group(1), m.group(2))

    # 9) /milk-sad/<chain>/<n>     -> /milk-sad/<chain>/1.html
    m = re.match(r'^milk-sad/([a-z0-9-]+)/([0-9]+)$', p)
    if m and _isfile('milk-sad/%s/1.html' % m.group(1)):
        return 'milk-sad/%s/1.html' % m.group(1)

    return None


class Handler(SimpleHTTPRequestHandler):
    # Match the .htaccess AddType directives so ES-module workers + WASM load.
    extensions_map = {
        **SimpleHTTPRequestHandler.extensions_map,
        '.js': 'application/javascript',
        '.mjs': 'application/javascript',
        '.wasm': 'application/wasm',
        '.json': 'application/json',
        '.svg': 'image/svg+xml',
    }

    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def _rewrite(self):
        parsed = urllib.parse.urlsplit(self.path)
        decoded = urllib.parse.unquote(parsed.path)
        target = resolve(decoded)
        if target is None:
            # ErrorDocument 404 /index.html  -> SPA-style fallback
            target = 'index.html' if _isfile('index.html') else decoded.strip('/')
        new = '/' + urllib.parse.quote(target)
        if parsed.query:
            new += '?' + parsed.query
        self.path = new

    def do_GET(self):
        self._rewrite()
        return super().do_GET()

    def do_HEAD(self):
        self._rewrite()
        return super().do_HEAD()


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8000
    httpd = ThreadingHTTPServer(('127.0.0.1', port), Handler)
    print('Serving %s' % ROOT)
    print('Clean-URL dev server (mirrors .htaccess) -> http://127.0.0.1:%d/' % port)
    print('Try: http://127.0.0.1:%d/bitcoin-puzzle/71/random' % port)
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print('\nStopped.')


if __name__ == '__main__':
    main()
